src/gui/IconDownloaderDialog.cpp

```diff
@@ -0,0 +1,186 @@
+    m_dataModel->setHorizontalHeaderLabels({"URL", "Status"});
+}
+
+IconDownloaderDialog::~IconDownloaderDialog()
+{
+    m_futureList.waitForFinished();
+}
+
+void IconDownloaderDialog::raiseWindow()
+{
+#ifdef Q_OS_MACOS
+    macUtils()->raiseOwnWindow();
+    Tools::wait(500);
+#endif
+}
+
+void IconDownloaderDialog::downloadFavicon(const QSharedPointer<Database>& database, Entry* entry)
+{
+    m_db = database;
+#ifdef WITH_XC_NETWORKING
+    QScopedPointer<DownloadIcon> downloader(new DownloadIcon());
+    connect(downloader.data(), SIGNAL(iconError()), this, SLOT(iconError()));
+    connect(downloader.data(), SIGNAL(fallbackNotEnabled()), this, SLOT(fallbackNotEnabled()));
+    connect(downloader.data(), SIGNAL(iconReceived(const QImage&, Entry*)), this, SLOT(iconReceived(const QImage&, Entry*)));
+    connect(m_ui->abortButton, SIGNAL(clicked()), downloader.data(), SLOT(abortRequest()));
+    connect(m_parent, SIGNAL(databaseLocked()), downloader.data(), SLOT(abortRequest()));
+    connect(m_parent, SIGNAL(databaseLocked()), this, SLOT(close()));
+
+    m_mutex.lock();
+    m_dataModel->appendRow(QList<QStandardItem*>() << new QStandardItem(entry->url())
+                                                   << new QStandardItem(tr("Loading")));
+    m_mutex.unlock();
+
+    downloader->downloadFavicon(entry);
+    QEventLoop loop;
+    connect(downloader.data(), SIGNAL(iconReceived(const QImage&, Entry*)), &loop, SLOT(quit()));
+    loop.exec();
+#endif
+}
+
+bool IconDownloaderDialog::downloadAllFavicons(const QSharedPointer<Database>& database, const QList<Entry*>& entries)
+{
+    m_db = database;
+
+    // Set progress bar
+    int maximum = 0;
+    for (const auto& e : entries) {
+        if (!e->url().isEmpty() && e->iconUuid().isNull()) {
+            ++maximum;
+        }
+    }
+    m_ui->progressBar->setMinimum(0);
+    m_ui->progressBar->setMaximum(maximum);
+
+    for (const auto& e : entries) {
+        if (!e->url().isEmpty() && e->iconUuid().isNull()) {
+            QFuture<void> fut = QtConcurrent::run(this, &IconDownloaderDialog::downloadFavicon, database, e);
+            m_futureList.addFuture(fut);
+        }
+    }
+
+    m_futureList.waitForFinished();
+    m_ui->abortButton->setEnabled(false);
+    return true;
+}
+
+void IconDownloaderDialog::iconReceived(const QImage& icon, Entry* entry)
+{
+    QMutexLocker locker(&m_mutex);
+    m_ui->progressBar->setValue(m_ui->progressBar->value() + 1);
+    m_ui->label->setText(tr("Downloading favicon %1 of %n…", 0, m_ui->progressBar->maximum()).arg(m_ui->progressBar->value()));
+
+    // Update the table
+    for (int i = 0; i < m_dataModel->rowCount(); ++i) {
+        if (m_dataModel->item(i, 0)->text() == entry->url()) {
+            m_dataModel->item(i, 1)->setText(icon.isNull() ? tr("Error") : tr("Done"));
+        }
+    }
+
+    if (!addCustomIcon(icon, entry)) {
+        emit messageEditEntry(tr("Custom icon already exists"), MessageWidget::Information);
```

Comment:
This would look better at the "Status" column. When there are a lot of icons being download at once, I have no idea which one this message is referring to.